### PR TITLE
Divyanshu | Test | Implementation of Zimbabwe VAT in GIDDH - Version 2

### PR DESCRIPTION
**Pull request by dvCodeWorld**
https://app.clickup.com/t/86cv0mmu4

### CODE REVIEW

File: apps/web-giddh/src/app/services/vat.service.ts [hunk 1]
```diff
@@ -226,4 +226,59 @@ export class VatService {
...
+     * @memberof VatService
+     */
+    public downloadVatLiabilityReport(request: VatReportRequest): Observable<BaseResponse<any, any>> {
+        this.companyUniqueName = this.generalService.companyUniqueName;
+        let url = this.config.apiUrl + VAT_API.DOWNLOAD_ZW_TRANSACTIONS_REPORT;
+        url = url?.replace(':companyUniqueName', encodeURIComponent(this.companyUniqueName));
+        url = url?.replace(':from', request.from);
+        url = url?.replace(':to', request.to);
+        url = url?.replace(':taxNumber', request.taxNumber);
+        url = url?.replace(':currencyCode', request.currencyCode ?? 'BWP');
```
Comment (dilpreetsinghofficial) [inline apps/web-giddh/src/app/services/vat.service.ts]: Remove static currency code

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    public isLoading: boolean;
+    /** This will hold the value out/in to open/close setting sidebar popup */
+    public asideGstSidebarMenuState: string = 'in';
+    /** Holds Current Currency Symbol for Zimbabwe report */
+    public vatReportCurrencySymbol: string = 'P';
+    /** True, if API is in progress */
+    public isTaxApiInProgress: boolean;
+    /** Holds Table Data  */
+    public vatLiabilityOverviewReport: any[];
+    public activeCompany: any = null;
```
Comment (dilpreetsinghofficial) [inline apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts:43]: Add documentation

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        this.initLiabilityReport();
+    }
+
+    /**
+    * VAT Obligations API Call
+    *
+    * @memberof LiabilityReportComponent
+    */
+    public getVatLiabilityReport(): void {
+        if (this.getFormControl('taxNumber').value) {
```
Comment (dilpreetsinghofficial) [inline apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts]:
```suggestion
        if (this.getFormControl('taxNumber')?.value) {
```

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+            this.isLoading = true;
+            this.vatService.getVatLiabilityReport(this.liabilityReportForm.value).pipe(takeUntil(this.destroyed$)).subscribe(response => {
+
+                this.isLoading = false;
+                if (response && response.status === "success" && response.body?.sections) {
+                    this.vatLiabilityOverviewReport = response.body.sections
```
Comment (dilpreetsinghofficial) [inline apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts]:
```suggestion
                    this.vatLiabilityOverviewReport = response.body.sections;
```

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        }
+    }
+
+    /**
+     * Redirect to liability detailed report list page
+     *
+     * @param {*} report
+     * @memberof LiabilityReportComponent
+     */
+    public viewDetailedReport(report): void {
```
Comment (dilpreetsinghofficial) [inline apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts]:
```suggestion
    public viewDetailedReport(report : any): void {
```

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+    /**
+    * Used to get and set form control value
+    *
+    * @param {string} control
+    * @returns {*}
+    * @memberof LiabilityReportComponent
+    */
+    public getFormControl(control: string): any {
+        return this.liabilityReportForm.get(control)
```
Comment (dilpreetsinghofficial) [inline apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts]:
```suggestion
        return this.liabilityReportForm.get(control);
```

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        }
+    }
+
+    /**
+     * Export Liability report based on currency and tax number
+     *
+     * @memberof LiabilityReportComponent
+     */
+    public downloadVatLiabilityReport(): void {
+        this.liabilityReportForm.value
```
Comment (dilpreetsinghofficial) [inline apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts]: ?

File: apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+    @Output() public currentBranchChange: EventEmitter<any> = new EventEmitter<any>();
+    /** Emits current Tax Number */
+    @Output() public currentTaxNumber: EventEmitter<any> = new EventEmitter<any>();
+    /** Emits get report to call api from parent component */
+    @Output() public getReport: EventEmitter<any> = new EventEmitter<any>();
+    /** Emits export/download report to call api from parent component */
+    @Output() public exportReport: EventEmitter<any> = new EventEmitter<any>();
+    /** Emits true when tax api call inprogress */
+    @Output() public isTaxApiInProgress: EventEmitter<any> = new EventEmitter<any>();
+
```
Comment (dilpreetsinghofficial) [inline apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts]: remove space

File: apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+        );
+    }
+
+    /**
+     * This will hide the datepicker
+     *
+     * @memberof VatReportFiltersComponent
+     */
+    public hideGiddhDatepicker(): void {
+        this.modalRef.hide();
```
Comment (dilpreetsinghofficial) [inline apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts]:
```suggestion
        this.modalRef?.hide();
```

File: apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+
+    /**
+     * This will get all the financial years of the company
+     *
+     * @memberof VatReportFiltersComponent
+     */
+    private getFinancialYears(): void {
+        this.settingsFinancialYearService.getFinancialYearLimits().pipe(takeUntil(this.destroyed$)).subscribe(res => {
+            if (res.body.startDate && res.body.endDate) {
+                this.createFinancialYearsList(res.body.startDate, res.body.endDate)
```
Comment (dilpreetsinghofficial) [inline apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts]:
```suggestion
                this.createFinancialYearsList(res.body.startDate, res.body.endDate);
```

File: apps/web-giddh/src/app/vat-report/liability-detailed-report/liability-detailed-report.component.ts [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    templateUrl: './liability-detailed-report.component.html',
+    styleUrls: ['./liability-detailed-report.component.scss']
+})
+export class LiabilityDetailedReportComponent implements OnInit, OnDestroy {
+    /** Observable to unsubscribe all the store listeners to avoid memory leaks */
+    private destroyed$: ReplaySubject<boolean> = new ReplaySubject(1);
+    /** Holds Vat Liability Detailed Report data */
+    public vatLiabilityDetailedReport: any = {};
+    /** Holds Vat Liability Detailed Report request object */
+    public vatLiabilityReportRequest: VatDetailedReportRequest = {
```
Comment (dilpreetsinghofficial) [inline apps/web-giddh/src/app/vat-report/liability-detailed-report/liability-detailed-report.component.ts:20]: check this

**Review by dilpreetsinghofficial — CHANGES_REQUESTED** [review]

File: apps/web-giddh/src/app/services/vat.service.ts (hunk 1 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/services/vat.service.ts]: Done

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts (hunk 2 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts:43]: Done

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts (hunk 3 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts]: Done

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts (hunk 4 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts]: Done

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts (hunk 5 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts]: Done

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts (hunk 6 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts]: Done

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts (hunk 7 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts]: Unused, and removed

File: apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts (hunk 8 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts]: Done

File: apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts (hunk 9 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts]: Done

File: apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts (hunk 10 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts]: Done

File: apps/web-giddh/src/app/vat-report/liability-detailed-report/liability-detailed-report.component.ts (hunk 11 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/liability-detailed-report/liability-detailed-report.component.ts:20]: Checked correct

Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/liability-detailed-report/liability-detailed-report.component.ts:20]: Done

**Review by dilpreetsinghofficial — APPROVED** [review]

File: apps/web-giddh/src/app/vat-report/liability-detailed-report/liability-detailed-report.component.ts [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    /**
+     * This will be use for component initialization
+     *
+     * @memberof LiabilityDetailedReportComponent
+     */
+    public ngOnInit(): void {
+        document.querySelector('body').classList.add('gst-sidebar-open');
+
+        this.route.queryParams.pipe(takeUntil(this.destroyed$)).subscribe(params => {
+            this.vatLiabilityReportRequest.from = params['from'];
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vat-report/liability-detailed-report/liability-detailed-report.component.ts]: add if condition to check params

File: apps/web-giddh/src/app/vat-report/liability-detailed-report/liability-detailed-report.component.ts [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    /**
+    * Lifecycle hook for destroy
+    *
+    * @memberof LiabilityDetailedReportComponent
+    */
+    public ngOnDestroy(): void {
+        this.destroyed$.next(true);
+        this.destroyed$.complete();
+        document.querySelector('body').classList.remove('gst-sidebar-open');
+        this.asideGstSidebarMenuState === 'out'
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vat-report/liability-detailed-report/liability-detailed-report.component.ts]:
```suggestion
        this.asideGstSidebarMenuState === 'out';
```

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+
+    /**
+    * Lifecycle hook for initialization
+    *
+    * @memberof LiabilityReportComponent
+    */
+    public ngOnInit(): void {
+
+        this.store.pipe(select(state => state.session.activeCompany), takeUntil(this.destroyed$)).subscribe(activeCompany => {
+            if (activeCompany && this.activeCompany?.uniqueName !== activeCompany.uniqueName) {
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts]:
```suggestion
            if (activeCompany) {
```

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.html [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+                                <ng-template #clickableItem>
+                                    <a href="javascript:;" (click)="viewDetailedReport(vatReport)"> 
+                                        {{vatReport?.section }}
+                                    </a>
+                                </ng-template>
+                            </td>
+                            <td width="20%" class="text-right">
+                                <amount-field 
+                                    class="d-inline-block" 
+                                    [amount]="vatReport?.vattotal"
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.html:56]: vattotal hi he key ka naam?

File: apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts [hunk 16]
```diff
@@ -0,0 +1,450 @@
...
+        this.isTaxApiInProgress.emit(true);
+        this.gstReconcileService.getTaxDetails().pipe(takeUntil(this.destroyed$)).subscribe(response => {
+            if (response && response.body) {
+                this.taxes = response.body.map(tax => ({
+                    label: tax,
+                    value: tax
+                }));
+            }
+            this.isTaxApiInProgress.emit(false);
+            this.taxNumber = this.taxes[0]?.value;
```
Comment (ravinderwalkover) [inline apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts]: move this and next line under above if condition

**Review by ravinderwalkover — CHANGES_REQUESTED** [review]

File: apps/web-giddh/src/app/vat-report/liability-detailed-report/liability-detailed-report.component.ts (hunk 12 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/liability-detailed-report/liability-detailed-report.component.ts]: Done

File: apps/web-giddh/src/app/vat-report/liability-detailed-report/liability-detailed-report.component.ts (hunk 13 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/liability-detailed-report/liability-detailed-report.component.ts]: Done

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts (hunk 14 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.ts]: Done

File: apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.html (hunk 15 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/liability-report/liability-report.component.html:56]: Yes,  checked

File: apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts (hunk 16 above)
Comment (dvCodeWorld) [reply apps/web-giddh/src/app/vat-report/vat-report-filters/vat-report-filters.component.ts]: Done

**Review by ravinderwalkover — APPROVED** [review]

**Pull request merged**